src/wp-content/themes/twentytwentyone/assets/js/interactivity.js

```diff
@@ -0,0 +1,206 @@
+/**
+ * WordPress dependencies
+ */
+import { store, getContext, getElement } from '@wordpress/interactivity';
+
+function checkClass( element, className ) {
+    if ( element.classList.contains( className ) ) {
+        return element;
+    }
+    if ( element.parentElement && element.parentElement.classList.contains( className ) ) {
+        return element.parentElement;
+    }
+    if ( element.parentElement.parentElement && element.parentElement.parentElement.classList.contains( className ) ) {
+        return element.parentElement.parentElement;
+    }
+    return null;
+}
+
+const { state, actions } = store( 'twentytwentyone', {
+    state: {
+        isPrimaryMenuOpen: false,
+        windowWidth: 0,
+        prevScroll: 0,
+        isDarkMode: false,
+        isDarkModeTogglerHidden: false,
+    },
+    actions: {
+        togglePrimaryMenu: () => {
+            state.isPrimaryMenuOpen = ! state.isPrimaryMenuOpen;
+        },
+
+        openPrimaryMenu: () => {
+            state.isPrimaryMenuOpen = true;
+        },
+
+        closePrimaryMenu: () => {
+            state.isPrimaryMenuOpen = false;
+        },
+
+        toggleDarkMode: () => {
+            state.isDarkMode = ! state.isDarkMode;
+            window.localStorage.setItem( 'twentytwentyoneDarkMode', state.isDarkMode ? 'yes' : 'no' );
```

Comment:
Not that this is wrong, but noting here that this can also be done reactively, which is a bit safer in cases where `state.isDarkMode` would be modified from a few different places:

```html
data-wp-watch="callbacks.storeDarkMode"
```

```js
callbacks: {
  storeDarkMode: () => {
    // This callback is triggered each time `state.isDarkMode` changes, no matter from where.
    window.localStorage.setItem( 'twentytwentyoneDarkMode', state.isDarkMode ? 'yes' : 'no' );
  }
}
```

Reply:
@luisherranz Follow up question: Wouldn't your suggestion call `window.localStorage.setItem` any time that _any_ of the store's `state` properties are updated? If not, how does it not only to look for `state.isDarkMode` changes?

Reply:
The callback only runs when the `state.isDarkMode` value changes because it's only subscribed to the `state.isDarkMode` signal. 

That's simply [how signals work](https://dev.to/ryansolid/building-a-reactive-library-from-scratch-1i0p): `data-wp-watch` is a computation and when you access a property (`state.isDarkMode`), before returning its value, it checks if it's inside a computation. If it is, it adds the computation to its internal graph of dependencies. When `state.isDarkMode` is updated, it invalidates all its computations, and the computation callbacks run again, updating their values/DOM.

In the Interactivity API, the interface is transparent (transparent reactive programming) because we are using Proxies on top of `state` and `context` to track their access/assignments.

_By the way, terms are a bit vague in the world of signals. Computations are sometimes called effects, reactions or observers, and signals are sometimes called observables or reactive variables. Oh, and now with Svelte 5, they are also called runes 😆 ._

Reply:
@luisherranz Thanks! Updated in https://github.com/WordPress/wordpress-develop/pull/5795/commits/b3af9e0ffa657095737f8f281706942c7ca4f8b3 (please ignore the whitespace changes)

Another follow up question: I have now two `data-wp-watch` callbacks for the same state property on the same HTML element. I think that makes sense, given that they serve different purposes, but I wanted to check whether there's any downside to this approach compared to putting the logic for both into the same callback, e.g. performance-wise.

Reply:
> but I wanted to check whether there's any downside to this approach compared to putting the logic for both into the same callback, e.g. performance-wise.

It's better to keep them separate because they will only be executed when they are strictly needed.

> The isDarkMode value should only be cached in localStorage if it was modified by the user

How does it work? When the user hasn't clicked, it follows `prefers-color-scheme`?

Reply:
@luisherranz
> How does it work? When the user hasn't clicked, it follows `prefers-color-scheme`?

Correct. So in that case the `prefers-color-scheme` should be relied on and not stored in local storage. Because otherwise, imagine the `prefers-color-scheme` value from the browser / OS changed, but it wouldn't be reflected due to the old value in local storage.

Local storage should only be used to persist the _explicit_ user choice.

Reply:
Can the user "unselect" the color preference to go back to `prefers-color-scheme`? So toggle between light/dark/`prefers-color-scheme`?

If so, I'd check the value of the setting in the watch callback and use `removeItem()` when the user has selected `prefers-color-scheme`, which should also be the initial value and therefore nothing should be saved on page load.

Would that make sense?

Reply:
@luisherranz That isn't possible. As soon as the user makes a choice in that UI, it's "permanent".

Reply:
Ok. So, even though the user can't go back to _system_ once they select light/dark, there are three possible values. So what I would do is to replace `state.isDarkMode = true | false`, which only stores two values, with `state.colorMode = 'system' | 'light' | 'dark'` to fully represent all the possibilities. 

The default value should be `system` (you can define it in PHP using `wp_interactivity_state` or in JS directly in the store definition). Then, I'd modify the watch callback to do something like this:

```js
callbacks: {
  storeDarkMode: () => {
    if ( state.colorMode !== 'system' ) {
      window.localStorage.setItem( 
        'twentytwentyoneDarkMode', 
        state.colorMode === 'dark' ? 'yes' : 'no' 
      );
    }
  }
}
```

Does that make sense?

I'm assuming you need to keep `twentytwentyoneDarkMode` for backward compatibility but if not you could use `twentytwentyoneColorMode` and store `light`/`dark`.

Reply:
@luisherranz That's a good idea, however with one modification necessary: We can't only store `system`, we would need to still know whether the system's value is light or dark. So we could use something like `system-light` and `system-dark`, but at that point I feel like we're overloading that property.

I went with two separate properties: `isDarkMode` remains as before, while a separate boolean property `isDarkModeManuallyOverwritten` (yes, very cumbersome name 😆) stores whether it's the system default or manually overwritten.

See https://github.com/WordPress/wordpress-develop/pull/5795/commits/4a35f3f24b81b2fecc5ec7a2d3d3f0446f9373e6

Reply:
Ok. But you should derive `isDarkMode`. Remember that all the state should have only one source of truth, so all the state that can be derived, should be derived.

Something like this:

```js
state: {
  colorPreference: window.localStorage.getItem( 'twentytwentyoneColorPreference' ) || "system",
  get isDarkMode() {
    if ( state.colorPreference === "system" )
      return window.matchMedia('(prefers-color-scheme: dark)').matches ? "dark" : "light";
    return state.colorPreference;
  }
},
actions: {
  toggleDarkMode() {
    state.colorPreference = state.colorPreference !== "dark" ? "dark" : "light";
  }
},
callbacks: {
  storeColorPreference() {
    if ( state.colorPreference !== 'system' ) {
      window.localStorage.setItem( 
        'twentytwentyoneColorPreference', 
        state.colorPreference
      );
    }
  }
}
```